### PR TITLE
feat(eslint-plugin): [switch-exhaustiveness-check] add an option to warn against a `default` case on an already exhaustive `switch`

### LINKED ISSUE #3616: [switch-exhaustiveness-check] ban the default case to enforce all cases are handled and prevent accidental mishandling of new cases
I frequently see code like this:

```ts
type MyUnion = 'a' | 'b';

declare const myUnion: MyUnion;

const f1 = () => (myUnion === 'a' ? 'a' : 'b');

f1('a') // 'a'
f1('b') // 'b'
```

The logic in `f1` makes an assumption: if `myUnion` is not `a`, it must be `b`.

Later on, someone might update the `MyUnion` type and this assumption will breakdown:

```diff
-type MyUnion = 'a' | 'b';
+type MyUnion = 'a' | 'b' | 'c';
```

The runtime behaviour is clearly incorrect, yet TypeScript will not error to remind us that we need to update the logic in `f1`:

```ts
f1('a') // 'a'
f1('b') // 'b'
f1('c') // 'b' ❌
```

This problem is not specific to the ternary operator but also `if` and `switch` statements:

```ts
const f2 = () => {
    if (myUnion === 'a') {
        return 'a';
    } else {
        return 'b';
    }
};

const f3 = () => {
    switch (myUnion) {
        case 'a':
            return 'a';
        default:
            return 'b';
    }
};
```

As we can see, it is not safe to make assumptions about the value that reaches the `else`/`default` case because it can change.

Instead we need to explicitly specify all cases:

```ts
import assertNever from 'assert-never';

type MyUnion = 'a' | 'b';

declare const myUnion: MyUnion;

const f2 = () => {
    if (myUnion === 'a') {
        return 'a';
    } else if (myUnion === 'b') {
        return 'b';
    } else {
        assertNever(myUnion);
    }
};

const f3 = () => {
    switch (myUnion) {
        case 'a':
            return 'a';
        case 'b':
            return 'b';
    }
};

const f3b = () => {
    switch (myUnion) {
        case 'a':
            return 'a';
        case 'b':
            return 'b';
        default:
            assertNever(myUnion);
    }
};
```

This way, when the type is eventually widened, TypeScript will generate a type error so we're reminded that we need to update our code:

```ts
import assertNever from 'assert-never';

type MyUnion = 'a' | 'b' | 'c';

declare const myUnion: MyUnion;

const f2 = () => {
    if (myUnion === 'a') {
        return 'a';
    } else if (myUnion === 'b') {
        return 'b';
    } else {
        // Argument of type 'string' is not assignable to parameter of type 'never'.
        assertNever(myUnion);
    }
};

// @noImplicitReturns: true
// Not all code paths return a value.
const f3 = () => {
    switch (myUnion) {
        case 'a':
            return 'a';
        case 'b':
            return 'b';
    }
};

const f3b = () => {
    switch (myUnion) {
        case 'a':
            return 'a';
        case 'b':
            return 'b';
        default:
            // Argument of type 'string' is not assignable to parameter of type 'never'.
            assertNever(myUnion);
    }
};
```

I would like to propose a rule that enforces this. The rule would report an error inside a ternary or `if`/`switch` statement if we're switching over a union type (except `boolean`) and we have a fallback case (`else`/`default`). The fix would be to explicitly specify all cases.

I'm really not sure what we would call it.

WDYT?

### PR DESCRIPTION
**Pull request by Zamiell**
## PR Checklist

- [X] Addresses an existing open issue: fixes #3616
- [X] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [X] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

I added a second check to the `switch-exhaustiveness-check` rule. The docs have been updated to say:

> Additionally, this also reports when a `switch` statement has a case for everything in a union and _also_ contains a `default` case. `default` cases in this situation are obviously superfluous, as they would contain dead code. But beyond being superfluous, these kind of `default` cases are actively harmful: if a new value is added to the switch statement union, the `default` statement would prevent this rule from alerting you that you need to handle the new case.

Currently, the rule has no options. Rather than adding a new option for this behavior, I decided that it should not be configurable, because I suspect that everyone will want to have this feature on by default and never turn it off.

## More Details

Currently, the `switch-exhaustiveness-check` rule works like this:

1. if the thing switching on is not a union, early return
2. if the switch statement has a default case, early return
3. if missing branches > 0, trigger the rule

I refactored the file such that the logic flow is now:

1. if the thing switching on is not a union, early return
2. gather metadata about the switch statement (which mainly includes the missing branches)
3. check `switchIsNotExhaustive` (fed with the metadata)
4. check `dangerousDefaultCase` (fed with the metadata)

### CODE REVIEW

**Comment by Josh-Cena** [comment]
Should we exempt the common case where the default case is a single throw statement?

**Comment by Zamiell** [comment]
Why would we exempt that? That's super dangerous for exactly the reasons that I put into the documentation.

**Comment by Zamiell** [comment]
The whole point of using TypeScript is that developers can uncover bugs in the code at compile-time instead end-users discovering bugs in the code at run-time. Finding at the switch statement error at run-time seems completely unacceptable.

**Review by bradzacher — CHANGES_REQUESTED** [review]
Not a full review - just a quick comment.
I think we will have to put this behind an option.
I know a lot of people also turn on the  `default-case` rule (even though they shouldn't really use both rules together).

And if we ship this as is then those users would now be unable to satisfy the linter (as this rule would force them to remove the `default`, then `default-case` would force them to re-add it).

There's a whole thought experiment here about how one could indicate to the rule whether or not a switch is meant to be exhaustive or not.. maybe that's a future extension for this rule. (eg a comment above the switch)
But for now we'll need to gate this behaviour

**Comment by Zamiell** [comment]
> I know a lot of people also turn on the default-case rule (even though they shouldn't really use both rules together).
> And if we ship this as is then those users would now be unable to satisfy the linter (as this rule would force them to remove the default, then default-case would force them to re-add it).

Can you expand more on why this is a requirement? Why do we care if a `typescript-eslint` rule conflicts with a core ESLint rule, especially when that core ESLint rule goes against TypeScript best-practices? Is there some design principle of `typescript-eslint` such that all of its rules should not conflict with base ESLint rules? I haven't heard about any such design principle until now.

Specifically, in this case I don't think the functionality should be behind an option because it prevents an actively harmful piece of code. I'll role-play the hypothetical you describe:

- In another routine day at the office, Alice updates the dependencies in her TypeScript project. However, she then notices some new errors from ESLint.
- Alice realizes that the new errors are because of an update to the `switch-exhaustiveness-check` rule.
- The lint rule advises her to delete the superfluous `default` case in a switch statement, so she does.
- However, frustratingly, doing that causes another ESLint rule to be triggered, the `default-case` rule.
- At this point, there is no quick fix, and Alice knows that she has to do a little bit more research to find out what is really going on:
  - Alice knows that the `switch-exhaustiveness-check` rule comes from the `typescript-eslint` team and that the `default` case rule comes from the core ESLint team. The former represents best-practices in the TypeScript ecosystem, and the latter is more generalized.
  - After reading the documentation for the `switch-exhaustiveness-check` rule, she understands more in depth what is going on - that the `default` case that the rule was advising her to delete was in actuality actively harmful, and that lint rule suggestion was amazingly good.
  - Furthermore, reading the documentation gives her some insight into why `default-case` is a terrible rule to turn on in TypeScript codebases.
- Alice now has the knowledge to safely turn off the `default-case` rule for the project (and all other TypeScript projects she manages).
- Alice is thankful for the helpful nod in the right direction from the `typescript-eslint` team, as she would have never realized that the `default-case` lint rule was actively harmful until now.

**Comment by Zamiell** [comment]
@JoshuaKGoldberg @bradzacher A month has passed with no reply, so pinging. Would love to get this merged without too much further delay.

**Comment by bradzacher** [comment]
This should be behind an option because not all codebases will want this behaviour.
As an example - at Meta we had a concept called "the safety window".
This window was the time during which a user may be on old JS code talking to a BE that is on newer code.
This concept existed because it was the time during which your code could break if it wasn't written in a way to be forwards-compatible.

As an example - imagine that you have an endpoint that returns a value of wither `0 | 1` - we'll call this "v1".
On Monday morning the user loads your web app and starts using it.
On Monday at lunch time you ship a change which causes the endpoint to to instead output `0 | 1 | 2`. We'll call this "v2".

If the user does not refresh their webpage - then they will be on "v1" of your JS code talking to "v2" of your backend.
If your code is written like this:
```ts
async function foo(): string {
  const value = await getValueFromBackend();
  switch (value) {
    case 0:
      return doSomething();

    case 1:
      return doSomeOtherThing();
  }
}
```

Then it means that your code will now behave in unexpected ways - most likely it will crash in some way, but it might also silently work correctly but do something fucky.

Either way - this is bad! So instead at Meta every `switch` statement _that deals with API data_ is enforced to be written with a `default` - ***EVEN IF IT IS EXHAUSTIVE***. This style means that you must explicitly handle the forwards-compatibility case - even if that handling means causing an explicit crash.

```ts
async function foo(): string {
  const value = await getValueFromBackend();
  switch (value) {
    case 0:
      return doSomething();

    case 1:
      return doSomeOtherThing();

    default:
      throw new Error("You're on outdated code - please refresh to update!");
  }
}
```

An explicit crash is better than an implicit one because it means you know for certain where and how your code will terminate - rather than potentially things continuing with and operating on corrupted data.
I.e. imagine if you were editing a google doc, added a new line and the page crashed. Then when you refreshed you found your entire doc was corrupted and lost forever because you'd been working with v1 JS code against a v2 backend. As a user you would be ***LIVID***. With an explicit error the app could have crashed out before it saved corrupted data and by refreshing your data would have been uncorrupted and recovered (maybe with some small data-loss).

By creating this functionality without an option - a company that follows the above code style would no longer be able to use the rule ***at all*** - it would have to be turned off for their entire codebase!

This is really bad!!!
Fun fact this is why [flow enums have special syntax to define "an enum with unknown members"](https://flow.org/en/docs/enums/defining-enums/#toc-flow-enums-with-unknown-members) - so that the API type generators can generate all API enums with this to allow lint rules to enforce switches are both exhaustive AND include a `default` case.

---

Another use cases is that not every switched value will be a union/enum value - sometimes they might just be `string` or `number`! For these cases a rule like `default-case` can be helpful to enforce that people are explicitly encoding their assumptions into their code.

However there's no way to make the rule just apply specifically to `switch`es that deal with non-union-type values - so it's on for everything.

So if this change isn't behind an option - that sort of user is now in the aforementioned unresolvable state! They want to use `default-case` for their non-union-type switches, but they cannot use it because this rule conflicts with it.

---

There are many usecases to consider! There's no harm in adding it behind an option.

**Comment by Zamiell** [comment]
Thanks for the explanation Brad. I will work on putting it behind an option today.

I am curious though: at Meta, after updating a V1 union to a V2 union, is there a way use tooling to find all the switch statements in a codebase that become out of date? Or do you just manually search for one of the enum members, and then hope that you got every one? I now understand the importance of throwing explicit errors to prevent undefined behavior, but I am also uneasy with "manually" updating switch statements (i.e. going back to life before the `switch-exhaustiveness-check` was invented`). Especially since years may have passed since the union update; no-one might not remember where all the switch statements are!

**Comment by JoshuaKGoldberg** [comment]
@Zamiell did you mean to re-request review from Brad? Unless we happen to come across PRs aby chance, we generally wait until an explicit re-request.

**Comment by Zamiell** [comment]
Well I was waiting for Brad to respond to my comment, which is separate from the review process. RE: my last comment, I would update the docs in the PR. Regardless, I'll explicitly request a review now.

**Comment by bradzacher** [comment]
> I am curious though: at Meta, after updating a V1 union to a V2 union, is there a way use tooling somehow to find all the switch statements in a codebase that become out of date and throw an error in CI?

For enums specifically - yes.
There is a flowlint rule which does this check.
You generally would prefix a `switch` with `// flowlint-next-line require-explicit-enum-switch-cases:error` and it'll enforce that even if the switch has a `default` - it also must have every case handled. Generally you don't have this rule on by default and instead opt-in to it on a case-by-case basis.

This is useful for "enums with unknown members" where you *have* to have a `default` case to handle the unknown state.

[flow playground](https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+IFkolXpUCWewUEAwhCQgRDH8wEH4hMnwROHlsNnw4KHwwSLAAC3wANyo4LFxscWQuHgMNZmwsiRSAWglaY1cq-hIAa2wJXNpG4Vxcdvdu3v7B0RxKUYMhKDBSqmbWwIq3eagoOrKSKgH0wtMMPznY7d2SfcoBiEZ-aG5G3Ix085AF-ZhsRoRehqUEiNMgSQHlSruBZxJrMcJwMhzAC+-EgGiCglEAAIAGLELHAAA6UCxpKxAEFeMSyViAEJUqBI4nEmALZRwaBYhr2ACMPIAFDBiMhccQAJQE6lkkgAdzgfRyWMF4slJJpZKyZFFECs5OQUvV6pYEiElCgAG5mWrDRotDp9dbDWSJDlKBAZVidh6AKLUCCUfliy2OpmMllsx4k7kSHkAJmVEBFeIgEqJjoA9OmsTAiDLcOkJIMFIX8zssSwAI5COAsRoKRj5sDyuvCEQguUKl4YMgkZAmN2UA1Y2Xy3JKoUp1VOrGa7Da3UO6c042mi1W6e2kL2ofTl1uj1erG+geB4M00Oh4mYkR0qhTsmUof0odWV-Ey+wCMcqNlCSxgAM-KYJQIq0lQqZDiOCpKsBEGOjSs63pQC47k6K5mmeTqIWByG0ouS6kuha7wWSmaehADRcjkkRYhgWKbtoEgzt2c4Ili1GUGwjSVqacApFi2D5IIWJwDAVFzteIkkAJ8jUXE-hCdm-q0SSkkdoqXRQO6JLmCIMwkEOF7hoskZcr+sYCsBoHgfepJQWOQE2Wm07YVQKEkYaRHBqhpIMduHnqnu7qetgPp+gGQaGe+xnspy0axvGVlIXBNJkTm7qloWOzyCW6RzjxNa-PWjbNte7ajjkXY9n24WQR2DmwbZ6qucheo+UavSrt5AV+RI+EEUFB6hUe4WnlFYZQDEIBCUC35BPk-5WAlADMS0gEiQA)

---

In TS - you're correct - using an "assert never" utility in the `default` case is a good option for ensuring exhaustiveness even with a `default`.

Nothing is assignable to `never` except `never`. You also can do a simple check like `thing satisfies never` to avoid the utility function if that's what you're into.

**Comment by Zamiell** [comment]
I went back and integrated the changes from #7880 now. It seems a little unfair that that #7880 was created in November, and mine was created in August, yet that one was merged first and caused me to re-do some work for this PR. Is there some sort of priority system that can prevent conflicts like this?

File: packages/eslint-plugin/tests/rules/switch-exhaustiveness-check.test.ts [hunk 1]
```diff
@@ -142,7 +142,7 @@ function test(value: Union): number {
   }
 }
     `,
-    // Switch contains default clause.
+    // Switch contains default clause
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/switch-exhaustiveness-check.test.ts:145]: [Proofreading] Unrelated changes. If you think we should standardize when/whether comments have punctutation / start with capital or lowercase letters then that'd be a separate issue to discuss from this PR. For now it'd be better (faster to review, cleaner diffs) to not make changes to code that wasn't otherwise touched.

File: packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts [hunk 2]
```diff
@@ -140,89 +301,38 @@ export default createRule<Options, MessageIds>({
...
-                },
-              },
-            ],
-          });
-        }
       }
     }
 
     return {
-      SwitchStatement: checkSwitchExhaustive,
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts:225]: [Praise] I like how you split this up into functions 🙂 much easier to read through the main logic flow than keeping it all inline would be. Nice!

File: packages/eslint-plugin/tests/rules/switch-exhaustiveness-check.test.ts (hunk 1 above)
Comment (Zamiell) [reply packages/eslint-plugin/tests/rules/switch-exhaustiveness-check.test.ts:145]: i looked through the comments and most of them did not have punctuation, so i standardized the style, which was only an extra 4 lines changed.
if you want i can remove those 4 changes from the PR?

File: packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md [hunk 3]
```diff
@@ -6,12 +6,55 @@ description: 'Require switch-case statements to be exhaustive.'
 >
 > See **https://typescript-eslint.io/rules/switch-exhaustiveness-check** for documentation.
 
-When working with union types or enums in TypeScript, it's common to want to write a `switch` statement intended to contain a `case` for each constituent (possible type in the union or the enum).
-However, if the union type or the enum changes, it's easy to forget to modify the cases to account for any new types.
+When working with union types or enums in TypeScript, it is common to want to write a `switch` statement intended to contain a `case` for each possible type in the union or the enum.
+However, if the union type or the enum changes, it is easy to forget to modify the cases to account for any new types.
```
[the remark this answers is not in the document]
Comment (Zamiell) [reply packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md:10]: i dont feel super strongly, do you want me to revert these lines?

File: packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md [hunk 4]
```diff
@@ -6,12 +6,55 @@ description: 'Require switch-case statements to be exhaustive.'
...
+
+    default: {
+      fruit satisfies never;
+      return 'unknownJuice';
+    }
+  }
+}
+```
+
+Doing this gives you the best of both worlds: explicit out-of-band value handling while still having the ability of TypeScript to alert to you all the places in your code-base where a new type constituent needs to be added. However, the downside is two-fold: there is no way for the `default-case` lint rule to distinguish between a `satisfies never` default case and some other kind of default case, so unsafe switch statements can still sneak into your codebase. Additionally, you have to maintain `default` cases everywhere, which makes things much more verbose.
```
[the remark this answers is not in the document]
Comment (Zamiell) [reply packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md:50]: I think it's important to include Brad's suggestion of `satisfies never` somewhere in there, can we add that to your more-terse version?

File: packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md:10]: If you don't mind that'd be better yeah.

File: packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts [hunk 5]
```diff
@@ -36,28 +63,159 @@ export default createRule<Options, MessageIds>({
...
+        missingBranchTypes,
+        symbolName,
+        defaultCase,
+      };
+    }
+
+    function checkSwitchNoUnionDefaultCase(
+      node: TSESTree.SwitchStatement,
+    ): void {
+      const hasDefault = node.cases.some(switchCase => switchCase.test == null);
```
[the remark this answers is not in the document]
Comment (Zamiell) [reply packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts]: > There are now two places where a default case is searched for
sorry about that, see [this commit](https://github.com/typescript-eslint/typescript-eslint/pull/7539/commits/cab680a68f589da9bf077b8bb3189b9383ca9b58) which adds `isUnion` and simplifies the code flow even further

File: packages/eslint-plugin/tests/rules/switch-exhaustiveness-check.test.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/switch-exhaustiveness-check.test.ts:145]: Yes please :)

File: packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts [hunk 6]
```diff
@@ -36,28 +63,159 @@ export default createRule<Options, MessageIds>({
...
+        // `default` case.
+        if (switchCase.test == null) {
+          continue;
+        }
+
+        const caseType = getConstrainedTypeAtLocation(
+          services,
+          switchCase.test,
+        );
+        caseTypes.add(caseType);
```
[the remark this answers is not in the document]
Comment (Zamiell) [reply packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts:127]: i don't 100% remember as this change is from some months ago, but glancing it it now I think I am trying to better satisfy SLR principle: https://midu.dev/single-line-responsability-haz-una-cosa-por-linea/

[the remark this answers is not in the document]
Comment (Zamiell) [reply packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts:127]: IME if prettier is splitting up the line into multiple lines, its a code smell that you are violating SLR and should refactor

File: packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts [hunk 7]
```diff
@@ -36,28 +54,139 @@ export default createRule<Options, MessageIds>({
...
+        isUnion: true,
+      };
+    }
+
+    function checkSwitchExhaustive(
+      node: TSESTree.SwitchStatement,
+      switchStatementMetadata: SwitchStatementMetadata,
+    ): void {
+      const { missingBranchTypes, symbolName, defaultCase } =
+        switchStatementMetadata;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts:155]: [Style] Is the reason for not destructuring inline so as to not make them break across lines? If so, have you considered changing the interface name to `SwitchMetadata`? The _"Statement"_ in there is arguably inferable from the context anyway...

```suggestion
      { missingBranchTypes, symbolName, defaultCase }: SwitchMetadata,
    ): void {
```

This is very much not a blocker at all 😄 just mentioning in case you like it.

**Review by JoshuaKGoldberg — DISMISSED** [review]
This is looking great, thanks so much for all your hard work on it! ❤️‍🔥 

Marking as approved from my end. Only a couple of comments, neither are particularly important to me. Just there if you find them useful.

I'll wait a bit for a review from @bradzacher since he also was looking at this. But if Brad doesn't have time, no worries, we can merge.

File: packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts (hunk 7 above)
Comment (Zamiell) [reply packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts]: i renamed it to switchMetadata now, thanks.

as for why I destructured it, it was for two reasons:
1) to avoid repetition, i.e. `if (x + y + z)` is easier to read than `if (switchMetadata.x + switchMetadata.y + switchMetadata.z)`
2) to avoid prettier breaking lines

Comment (Zamiell) [reply packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts]: i guess another benefit is that it prevents mutation, since we are working on a non-read-only version of the object, and thus we should make copies of the values instead of working with them on the object directly, if we did that they we would be more likely to accidently mutate it

File: packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md [hunk 8]
```diff
@@ -6,12 +6,51 @@ description: 'Require switch-case statements to be exhaustive.'
...
-When working with union types or enums in TypeScript, it's common to want to write a `switch` statement intended to contain a `case` for each constituent (possible type in the union or the enum).
+When working with union types or enums in TypeScript, it's common to want to write a `switch` statement intended to contain a `case` for each possible type in the union or the enum.
 However, if the union type or the enum changes, it's easy to forget to modify the cases to account for any new types.
 
 This rule reports when a `switch` statement over a value typed as a union of literals or as an enum is missing a case for any of those literal types and does not have a `default` clause.
 
-There is also an option to check the exhaustiveness of switches on non-union types by requiring a default clause.
+## Options
+
+### `"allowDefaultCaseForExhaustiveSwitch"`
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md:16]: Filed #8154 as a followup 👍
